st2common/st2common/router.py

```diff
@@ -0,0 +1,386 @@
+                    declaration, options = statement.copy().popitem()
+                    definition = security_definitions[declaration]
+
+                    if definition['type'] == 'apiKey':
+                        if definition['in'] == 'header':
+                            token = req.headers.get(definition['name'])
+                        elif definition['in'] == 'query':
+                            token = req.GET.get(definition['name'])
+                        else:
+                            token = None
+
+                        if token:
+                            if auth_resp:
+                                raise auth_exc.MultipleAuthSourcesError(
+                                    'Only one of Token or API key expected.')
+
+                            auth_func = op_resolver(definition['x-operationId'])
+                            auth_resp = auth_func(token)
+
+                            context['user'] = User.get_by_name(auth_resp.user)
+
+                if 'user' not in context:
+                    raise auth_exc.NoAuthSourceProvidedError('One of Token or API key required.')
+            except (auth_exc.NoAuthSourceProvidedError,
+                    auth_exc.MultipleAuthSourcesError) as e:
+                LOG.error(str(e))
+                return abort_unauthorized(str(e))
+            except auth_exc.TokenNotProvidedError as e:
+                LOG.exception('Token is not provided.')
+                return abort_unauthorized(str(e))
+            except auth_exc.TokenNotFoundError as e:
+                LOG.exception('Token is not found.')
+                return abort_unauthorized(str(e))
+            except auth_exc.TokenExpiredError as e:
+                LOG.exception('Token has expired.')
+                return abort_unauthorized(str(e))
+            except auth_exc.ApiKeyNotProvidedError as e:
+                LOG.exception('API key is not provided.')
+                return abort_unauthorized(str(e))
+            except auth_exc.ApiKeyNotFoundError as e:
+                LOG.exception('API key is not found.')
+                return abort_unauthorized(str(e))
+            except auth_exc.ApiKeyDisabledError as e:
+                LOG.exception('API key is disabled.')
+                return abort_unauthorized(str(e))
+
+        if cfg.CONF.rbac.enable:
+            user_db = context['user']
+
+            permission_type = endpoint.get('x-permissions', None)
+            if permission_type:
+                resolver = resolvers.get_resolver_for_permission_type(permission_type)
+                has_permission = resolver.user_has_permission(user_db, permission_type)
+
+                if not has_permission:
+                    raise rbac_exc.ResourceTypeAccessDeniedError(user_db,
+                                                                 permission_type)
+
+        # Collect parameters
+        kw = {}
+        for param in endpoint.get('parameters', []) + endpoint.get('x-parameters', []):
```

Comment:
Quite some "magic" (aka new `jsexpose` in some ways) :P

It's definitively an improvement over `jsexpose` because now at least it uses global definitions, but I it's still a lot going on here so I would also prefer some standalone explicit tests for this functionality if we don't have some already.